### PR TITLE
Add search space for ray.tune

### PR DESCRIPTION
**Pull request by mberr**
This PR adds a method to construct a search space for [ray.tune](https://docs.ray.io/en/master/tune/index.html)

### CODE REVIEW

File: src/class_resolver/api.py [hunk 1]
```diff
@@ -118,6 +118,32 @@ def get_option(self, *flags: str, default: Optional[str] = None, **kwargs):
...
+
+                You can do so by selecting the appropriate install option for the package
+
+                    pip install class-resolver[ray]
+
+                or by manually installing ray tune
+
+                    pip install ray[tune]
+                """,
+            ))
```
Comment (cthoyt) [inline src/class_resolver/api.py:138]: If you do this, it won't show the hierarchy of exceptions raised during catching other exceptions
```suggestion
            )) from None
```

File: src/class_resolver/api.py [hunk 2]
```diff
@@ -118,6 +118,32 @@ def get_option(self, *flags: str, default: Optional[str] = None, **kwargs):
...
+        if kwargs_search_space is None:
+            return ray.tune.choice(sorted(self.lookup_dict.keys()))
```
Comment (cthoyt) [inline src/class_resolver/api.py:140]: this should be a bit nicer looking with that other PR merged in

File: src/class_resolver/api.py [hunk 3]
```diff
@@ -118,6 +118,32 @@ def get_option(self, *flags: str, default: Optional[str] = None, **kwargs):
             **kwargs,
         )
 
+    def ray_tune_search_space(self, kwargs_search_space: Optional[Mapping[str, Any]] = None):
+        """Return a search space for ray.tune."""
```
Comment (cthoyt) [inline src/class_resolver/api.py:122]: this needs some docs and examples

File: src/class_resolver/api.py [hunk 4]
```diff
@@ -118,6 +119,32 @@ def get_option(self, *flags: str, default: Optional[str] = None, **kwargs):
...
+                or by manually installing ray tune
+
+                    pip install ray[tune]
+                """,
+            )) from None
+        if kwargs_search_space is None:
+            return ray.tune.choice(sorted(self.lookup_dict.keys()))
+
+        return dict(
+            query=ray.tune.choice(sorted(self.lookup_dict.keys())),
```
Comment (cthoyt) [inline src/class_resolver/api.py]: you use this same line twice but just return different ways, why not save it as a variable before the first `if kwargs_search_space is None:` check

File: src/class_resolver/api.py (hunk 2 above)
Comment (mberr) [reply src/class_resolver/api.py]: c4450fe

File: src/class_resolver/api.py (hunk 4 above)
Comment (mberr) [reply src/class_resolver/api.py]: c67c5bf

File: src/class_resolver/api.py (hunk 3 above)
Comment (mberr) [reply src/class_resolver/api.py:140]: there is a small example in 
eb15dbd

File: tests/test_api.py [hunk 5]
```diff
@@ -55,3 +56,49 @@ def test_passthrough(self):
...
+        query = to_resolve
+        kwargs = None
+    resolver = Resolver([A, B, C], base=Base)
+    instance = resolver.make(query, pos_kwargs=kwargs)
+    if instance.name == "charlie":
+        mean_loss = 1.0
+    else:
+        mean_loss = 2.0
+
+    import ray.tune
```
Comment (cthoyt) [inline tests/test_api.py:78]: I added `ray` as a testing extra, so you can assume that it's importable in tests

File: tests/test_api.py [hunk 6]
```diff
@@ -55,3 +56,49 @@ def test_passthrough(self):
...
+    ray.tune.report(mean_loss=mean_loss)
+
+
+def test_ray():
```
Comment (cthoyt) [inline tests/test_api.py:81]: this example has a lot of code duplication - why don't you put it inside a unittest TestCase so you can save the resolver as a state variable and make dummy_training_function also a function of the class

Comment (mberr) [reply tests/test_api.py:81]: With this modification, I got the following error

```python-traceback
Error
Traceback (most recent call last):
  File "/usr/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/usr/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "../class-resolver/tests/test_api.py", line 64, in test_ray
    analysis = ray.tune.run(
  File "../venv/class-resolver/lib/python3.8/site-packages/ray/tune/tune.py", line 306, in run
    experiments[i] = Experiment(
  File "../venv/class-resolver/lib/python3.8/site-packages/ray/tune/experiment.py", line 149, in __init__
    self._run_identifier = Experiment.register_if_needed(run)
  File "../venv/class-resolver/lib/python3.8/site-packages/ray/tune/experiment.py", line 301, in register_if_needed
    raise type(e)(str(e) + " " + extra_msg) from None
_pickle.PicklingError: ray.cloudpickle.dumps(<class 'ray.tune.function_runner.wrap_function.<locals>.ImplicitFunc'>) failed.
To check which non-serializable variables are captured in scope, re-run the ray script with 'RAY_PICKLE_VERBOSE_DEBUG=1'. Other options: 
-Try reproducing the issue by calling `pickle.dumps(trainable)`. 
-If the error is typing-related, try removing the type annotations and try again.
```

Comment (mberr) [reply tests/test_api.py:81]: I suppose it is connected to ray, which creates new processes for the experiments.

Comment (mberr) [reply tests/test_api.py]: With [ray.tune.with_parameters](https://docs.ray.io/en/master/tune/api_docs/trainable.html?highlight=with_parameters#ray.tune.with_parameters) and a static method it works, except that I get some redis warnings now :sweat_smile: 
```
../venv/class-resolver/lib/python3.8/site-packages/ray/node.py:827: ResourceWarning: unclosed file <_io.TextIOWrapper name='/tmp/ray/session_2021-03-31_12-29-09_243337_20619/logs/redis-shard_0.err' mode='a' encoding='utf-8'>
  self.start_redis()
ResourceWarning: Enable tracemalloc to get the object allocation traceback
../venv/class-resolver/lib/python3.8/site-packages/ray/node.py:827: ResourceWarning: unclosed file <_io.TextIOWrapper name='/tmp/ray/session_2021-03-31_12-29-09_243337_20619/logs/redis-shard_0.out' mode='a' encoding='utf-8'>
  self.start_redis()
```

File: src/class_resolver/api.py [hunk 7]
```diff
@@ -135,6 +136,65 @@ def classes(self) -> Set[Type[X]]:
         """Return the available classes."""
         return set(self.lookup_dict.values())
 
+    def ray_tune_search_space(self, kwargs_search_space: Optional[Mapping[str, Any]] = None):
+        """Return a search space for ray.tune.
+
+        ray.tune is a package for distributed hyperparameter optimization. The search space for this search is defined
+        as a (nested) dictionary, which can contain special values `tune.{choice,uniform,...}`. For these values, the
```
Comment (mberr) [inline src/class_resolver/api.py:143]: @cthoyt here is something broken with the docstring formatting, What would be the correct way for inline code?

File: src/class_resolver/api.py [hunk 8]
```diff
@@ -135,6 +136,65 @@ def classes(self) -> Set[Type[X]]:
...
+        search algorithm will sample a specific configuration.
+
+        This method can be used to create a tune.choice sampler for the choices available to the resolver. By default,
+        this is equivalent to
+
+            ray.tune.choice(self.options)
```
Comment (mberr) [inline src/class_resolver/api.py:149]: this should be a code block

Comment (cthoyt) [reply src/class_resolver/api.py:151]: I added `.. code-block:: python` for you

**Review by cthoyt — APPROVED** [review]
🚢 :shipit: 🚢

**Comment by mberr** [comment]
@cthoyt ready to merge from my side

**Review by cthoyt — APPROVED** [review]

**Pull request merged**